File: Data Structures & Algorithms/house-robber-ii/submission-11.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {  
        if (nums.length === 1) return nums[0]
        const dfs = (i, flag, memo = {}) => { 
            if (i in memo) return memo[i]
            if (i >= nums.length || (flag && i === nums.length - 1)) return 0 

            const res =Math.max(nums[i] + dfs(i + 2, flag, memo), dfs(i + 1, flag, memo)) 
            memo[i] = res 
            return res
        } 

        return Math.max(dfs(0, true), dfs(1, false))
    }
}

File: Data Structures & Algorithms/partition-labels/submission-11.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        end = defaultdict(int) 

        for i, c in enumerate(s): 
            end[c] = i 

        curMax = 0 
        res = [] 
        l = 0  
        
        for i, c in enumerate(s): 
            curMax = max(end[c], curMax) 

            if curMax == i: 
                res.append(i - l + 1) 
                l = i + 1 

        return res